cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -261,58 +322,149 @@ private void restartIfNecessary(int podId, RestartContext restartContext)
+                log.debug("{}: Waiting for non-restarted pod {} to become ready", reconciliation, podId);
+                await(isReady(namespace, KafkaCluster.kafkaPodName(cluster, podId)), operationTimeoutMs, TimeUnit.MILLISECONDS, e -> new FatalProblem("Error while waiting for non-restarted pod " + podName(podId) + " to become ready", e));
+                log.debug("{}: Pod {} is now ready", reconciliation, podId);
+            }
+        } catch (ForceableProblem e) {
+            if (restartContext.backOff.done() || e.forceNow) {
+                log.warn("{}: Pod {} will be force-rolled", reconciliation, podName(podId));
+                restartAndAwaitReadiness(pod, operationTimeoutMs, TimeUnit.MILLISECONDS);
+            } else {
+                throw e;
+            }
+        } finally {
+            if (restartPlan != null) {
+                restartPlan.closeLoggingAnyError();
             }
-        } else {
-            // By testing even pods which don't need restart for readiness we prevent successive reconciliations
-            // from taking out a pod each time (due, e.g. to a configuration error).
-            // We rely on Kube to try restarting such pods.
-            log.debug("Pod {} does not need to be restarted", podId);
-            log.debug("Waiting for non-restarted pod {} to become ready", podId);
-            await(isReady(namespace, KafkaCluster.kafkaPodName(cluster, podId)), operationTimeoutMs, TimeUnit.MILLISECONDS, e -> new FatalProblem("Error while waiting for non-restarted pod " + podName(podId) + " to become ready", e));
-            log.debug("Pod {} is now ready", podId);
         }
     }
 
-    private void closeLoggingAnyError(Admin adminClient) {
-        if (adminClient != null) {
+    private RestartPlan restartPlan(int podId, Pod pod) throws ForceableProblem, InterruptedException, FatalProblem {
+        List<String> reasonToRestartPod = podNeedsRestart.apply(pod);
+        // Unless the annotation is present, check the pod is at least ready.
+        boolean needsRestart = reasonToRestartPod != null && !reasonToRestartPod.isEmpty();
+        KafkaBrokerConfigurationDiff diff = null;
+        boolean needsReconfig = false;
+        Admin adminClient = null;
+        if (!needsRestart) {
+            /*
+            First time we do the restart, the pod is Unschedulable, the post-restart livesness check fails with a timeout and the reconciliation ends
+              - lastProbeTime: null
+                lastTransitionTime: "2020-06-19T10:25:54Z"
+                message: '0/1 nodes are available: 1 Insufficient cpu.'
+                reason: Unschedulable
+                status: "False"
+                type: PodScheduled
+
+            Next reconciliation there's no reason to restart, so we try to determine whether we need to reconfigure.
+            We get CE. If this were not fatal we'd try the next broker (TEST FAIL).
```

Comment:
CE? ChangeException?

Reply:
Most probably. @tombentley ^

Reply:
CE=ConfigException. The problem we found was: 

* If we treated a ConfigException from the `AdminClient` as fatal then we weren't able to proceed with the restarting in the case where a running Pod is just not responding
* If we deleted a Pod and the new one was not schedulable and the resulting ConfigException was not fatal then we would end up taking down the cluster, one pod at a time, because

The solution I used here was to defer opening the AdminClient in the case where we already know we have reasons to restart (which includes the not schedulable test case). This meant there were two different code paths where we opened the AC and I could treat the exception differently in each. This was enough to pass the test case.

But thinking about it now, I realise that's flawed. The test causes the non-schedulability of a Pod by changing the CPU requirement to something unfulfillable. That requires a rolling restart because the Pod is stale wrt the STS pod template. But there are other causes of pods being unschedulable which don't result in stale pods, and those would go via the deferred AC creation path, and thus treat CE as non-fatal.

I think the correct way to solve it would be:

* Look at the Pod status prior to deciding to open an AC. Unschedulable => fatal exception (which will exit the reconciliation)
* We can then remove the deferred AC creation, since it's a complication we don't need.

I feel stupid, because it's blindingly obvious when I put it like this :disappointed: @stanlyDoge is this a change you're happy to make (given this very brief description)?